libs/isograph-react/src/core/optimisticProxy.ts

```diff
@@ -0,0 +1,411 @@
+import {
+  callSubscriptions,
+  insertEmptySetIfMissing,
+  type EncounteredIds,
+} from './cache';
+import type {
+  BaseStoreLayerData,
+  IsographEnvironment,
+  StoreLayerData,
+  StoreLink,
+  StoreRecord,
+} from './IsographEnvironment';
+
+export function getOrInsertRecord(
+  dataLayer: StoreLayerData,
+  link: StoreLink,
+): StoreRecord {
+  const recordsById = (dataLayer[link.__typename] ??= {});
+  return (recordsById[link.__link] ??= {});
+}
+
+export function readOptimisticRecord(
+  storeLayer: StoreLayer,
+  link: StoreLink,
+): StoreRecord {
+  return new Proxy<StoreRecord>(
+    {},
+    {
+      get(_, p) {
+        let node: StoreLayer | null = storeLayer;
+
+        while (node !== null) {
+          const storeRecord = node.data[link.__typename]?.[link.__link];
+          if (storeRecord != null) {
+            const value = Reflect.get(storeRecord, p);
+            if (value !== undefined) {
+              return value;
+            }
+          }
+          node = node.parentStoreLayer;
+        }
+      },
+      has(_, p) {
+        let node: StoreLayer | null = storeLayer;
+
+        while (node !== null) {
+          const storeRecord = node.data[link.__typename]?.[link.__link];
+          if (storeRecord != null) {
+            const value = Reflect.has(storeRecord, p);
+            if (value) {
+              return true;
+            }
+          }
+          node = node.parentStoreLayer;
+        }
+        return false;
+      },
+      set(_, p, newValue) {
```

Comment:
Maybe we should have two functions: `getMutableStoreRecordProxy` and `getStoreRecordProxy`, and call `getStoreRecordProxy` from `read` and `getMutable` from `cache`?

Because right now, the setter is defined but not valid in `read`, since the childmost store layer could be an optimistic layer or a startUpdate layer.